### ISSUE TITLE
ToDo list

### ISSUE THREAD

**Issue by jarun** [jarun/nnn#110]
Rolled from #58.

Feel free to suggest new features to add to the list below. Some of the features may still be master-only.

- [ ] support <kbd>TAB</kbd>-completion for scripts in PATH (see #127)
- [ ] view and access network drives on the LAN (sample script)
- [ ] ssh and scp support ([script with sshfs](https://www.digitalocean.com/community/tutorials/how-to-use-sshfs-to-mount-remote-file-systems-over-ssh))
- [x] retire `nlay` (no more used by `nnn`)
- [x] integrate static analysis on Travis
- [x] list currently copied files
- [x] option to disable directory auto-select
- [x] use `$TMPDIR` if available
- [x] support terminal locking on OS X (bashlock) and BSD (lock)
- [x] support `patool`
- [x] run any number of scripts
- [x] prefer VISUAL over EDITOR
- [x] add a keybind to run a custom script (see #79)
- [x] integer `get_coolsize()` (thanks @arsv)
- [x] batch rename, move, delete files with vidir integration
- [x] copy file paths when X is unavailable
- [x] support copying multiple file names to copier
- [x] add open with... shortcut
- [x] add option to open bookmark directly
- [x] optimize memory usage and algorithms for file name handling
- [x] list and extract archives
- [x] show current selection number along with total entries in dir
- [x] support `NNN_NOWAIT` to unblock nnn (DE specific)
- [x] file and directory creation
- [x] change to light mode if number of cols < 35
- [x] in-place file rename
- [x] watch out for current dir changes (if not in filter or du mode)
- [x] pin and jump to a directory
- [x] show vol capacity and vol free in help
- [x] option to use exiftool instead of mediainfo
- [x] option to show dirs in custom color (default: enabled in blue)
- [x] show dir content size and number of files in du mode
- [x] support bookmarks string
- [x] add option to support opening text files in EDITOR
- [x] use `xdg-open` on Linux and `open(1)` on OS X as desktop opener
- [x] support multiple apps in `nlay`
- [x] use `vlock` as default terminal screensaver
- [x] *navigate-as-you-type* mode
- [x] subtree search (using gnome-search-tool or catfish)
- [x] support custom nlay
- [x] *search-as-you-type* at filter prompt
- [x] RPM packages for RHEL/CentOS (thanks @fennm)
- [x] ~support cd ..... (with . as PWD, stops at /)~ (feature removed)
- [x] jump to initial directory (where nnn was started)
- [x] `nlay` - a shell script for actions like open desktop search utility, run screensaver
- [x] show file details, mediainfo output, help in `less`
- [x] show media information (needs mediainfo)
- [x] ~support readline at chdir prompt~ (feature removed)
- [x] use `Q` to quit to current dir (needs external shell integration)
- [x] use `^Q` to quit (even from filter mode)
- [x] UTF-8 support
- [x] add option to start in disk usage analyzer mode
- [x] open current dir in desktop file manager
- [x] disk usage analyzer more
- [x] show unprivileged free blocks in du mode
- [x] shortcut `-` to jump to last directory
- [x] ~resolve `~` (HOME) at chdir prompt~ (feature removed)
- [x] option to show complete stats of just the current entry (name, file type, human readable permission string, byte size... all `lstat()` details)
- [x] update noice man page
- [x] sort entries by size
- [x] support file path copier script (`NNN_COPIER`)
- [x] support detail view (default)
  - [x] file type
  - [x] modification time
  - [x] human-readable file size
  - [x] current item in reverse video
  - [x] full name of currently selected file in bottom "bar"
- [x] roll navigation with `Up` and `Down` at the first and last entries
- [x] remove restriction on visiting parent dir when opened with a relative dir path
- [x] case insensitive alphabetic content listing
- [x] generate optimized binary
- [ ] ~show total number of files in du mode~ (apparently it has to be counted in nftw callback, with adjustments like file type check, hardlink check, which means unreasonable overhead)
- [ ] ~threaded/multicore nftw(3)~ (overkill, we are already very fast)
- [ ] ~use scandir instead of readdir~ (we are using custom alphanum sort, see https://github.com/jarun/nnn/commit/691291245a88d1c2277d8900a13f41b00685a72c)
- [ ] ~modify packagecore.yml and probably Makefile to generate OpenSUSE Leap package~

Anything else which would add value (please discuss in this thread)

**Comment by jtgd** [comment]
- [ ] Move cursor down after ^K (to easily select multiple items) possibly with config option.

**Comment by jtgd** [comment]
Arun Prakash Jana <notifications@github.com> writes:

> User may want to select files which are not consecutive. ^K
> cherry-picks the files (in the next release, you would be able to use
> space too).

If they are picking non-consecutive files then they have to move the
cursor anyway, but in the consecutive case (more common) it would be
faster.

> So this can't be supported.

Then you could make it a config option, but that's ok, if you are
against it. Maybe I'll just fork a personal copy. :)

--
Jay Dresser

**Comment by jarun** [comment]
> but in the consecutive case (more common) it would be faster.

I am saying in case of consecutive files you can press `^Y` at the first and last files to copy all of them.

**Comment by InfinityPhase** [comment]
- [ ] Support for network drives (Such as SAMBA)

**Comment by InfinityPhase** [comment]
It would be convenient if (like Dolphin and the like) there was support to both view and access network drives on the LAN. I'm not sure what is involved in that, so I don't know the feasibility.

**Comment by jarun** [comment]
Adding it to the list.

**Comment by jarun** [comment]
@InfinityPhase it may be possible to use scripts to list networks, and mount them locally using smbclient from the cmdline.

refs:

- https://security.stackexchange.com/questions/36198/how-to-find-live-hosts-on-my-network
- https://unix.stackexchange.com/questions/8118/how-to-find-what-other-machines-are-connected-to-the-local-network
- https://superuser.com/questions/311019/display-list-of-computers-on-a-lan-in-linux

The next `nnn` release will have support for any number of user scripts.

**Comment by jarun** [comment]
@tom-i you can use custom scripts to do ssh mount (https://www.digitalocean.com/community/tutorials/how-to-use-sshfs-to-mount-remote-file-systems-over-ssh) or any kind of remote fs mount.

**Comment by leorosa** [comment]
How about:
- [ ] assign keys to different scripts, e.g. "@1" to run one script, "@2" to run another script, etc.;
- [ ] set bookmarks on-the-fly, in a way similar to which vim uses to define 'marks' in a file (m+letter to mark a position, '+letter to return to that position); this would be better than using NNN_BMS and the 'pinned' directory, which is already very useful;

**Comment by jarun** [comment]
> assign keys to different scripts, e.g. "@1" to run one script, "@2" to run another script, etc.;

I wouldn't consider a fixed mapping of keys. The current arrangement works pretty well.

> which is already very useful

I don't find a compelling need to change the way these things work today.

I think both your proposals are to modify things those work pretty well already. Please let me know if you have ideas which can reduce the amount of configuration. ;)

**Comment by leorosa** [comment]
Thanks for your consideration. The thing with marking directories at run time is that it would work without the need to set NNN_BMS -- but it would only be useful if the sessions was saved somewhere. But I agree, its not compelling.

Regarding the scripts... you already have "R" as a fixed key map, and you already read various scripts... you could map "R" to run a script, if only one is found, or "R1", "R2", etc., if several scripts are found. Without new configuration. I use "R" to call a script which move files in the clipboard to the current dir. This mechanism would be useful to enable users to also copy files, delete files, etc.

Again, thanks for your consideration. This program is already very useful for me as is :)

**Comment by jarun** [comment]
`R` can already be used to invoke multiple scripts through `NNN_MULTISCRIPT`. I am not inclined to add multiple key assignments.

**Comment by Nexion** [comment]
It would be nice to have symlinked directories sorted among the other directories. Currently all symlinks are sorted among other files.

**Comment by jarun** [comment]
The behaviour is same as `ls`:

```
~$ mkdir test
~$ ln -s test testlink
~$ ls -l
total 276
drwxrwxr-x  2 vaio vaio   4096 Aug  4 16:55 Desktop/
drwxr-xr-x  8 vaio vaio   4096 Aug 14 07:46 Documents/
drwxr-xr-x  5 vaio vaio  20480 Aug 15 11:23 Downloads/
drwxrwxr-x  4 vaio vaio   4096 Mar  8 07:54 Dropbox/
drwxrwxr-x 28 vaio vaio   4096 Aug  1 10:06 GitHub/
drwxrwxr-x 12 vaio vaio   4096 Jul 27 22:00 Music/
drwxr-xr-x  2 vaio vaio  12288 Aug 15 03:12 Pictures/
drwxrwxr-x  2 vaio vaio   4096 Aug 15 21:42 test/
drwxr-xr-x  4 vaio vaio  12288 Aug  8 02:31 Videos/
-rw-rw-r--  1 vaio vaio 186368 Aug 15 08:43 bookmarks.db
lrwxrwxrwx  1 vaio vaio      4 Aug 15 21:42 testlink -> test/
```

Also, technically a symlink is not a directory. Detecting if the target is a file or a dir involves extra processing.

**Comment by Nexion** [comment]
The default behavior of my `ls` (coreutils-8.29) is to sort all the files alphabetically, regardless if it's a directory, a symlink, an ordinary file or even a named pipe:
```
$  ls -l
total 8
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:31 anime
-rw-r--r-- 1 nexion nexion  0 Aug 15 19:32 av.conf
-rw-r--r-- 1 nexion nexion  0 Aug 15 19:31 bookmarks.txt
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:31 documents
lrwxrwxrwx 1 nexion nexion 23 Aug 15 19:34 downloads -> /mnt/hdd/downloads
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:34 library
lrwxrwxrwx 1 nexion nexion 19 Aug 15 19:34 music -> /mnt/hdd/music
prw-r--r-- 1 nexion nexion  0 Aug 15 19:37 namedpipe
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:35 something_else
```

However, if you add the option `--group-directories-first` it will list directories first, like in your output. And if you add the option `--dereference` it will treat symlinks like the files they actually are referencing to:

```
$ ls -l --group-directories-first
total 8
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:31 anime
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:31 documents
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:34 library
drwxr-xr-x 1 nexion nexion  0 Aug 15 19:35 something_else
-rw-r--r-- 1 nexion nexion  0 Aug 15 19:32 av.conf
-rw-r--r-- 1 nexion nexion  0 Aug 15 19:31 bookmarks.txt
lrwxrwxrwx 1 nexion nexion 23 Aug 15 19:34 downloads -> /mnt/hdd/downloads
lrwxrwxrwx 1 nexion nexion 19 Aug 15 19:34 music -> /mnt/hdd/music
prw-r--r-- 1 nexion nexion  0 Aug 15 19:37 namedpipe

$  ls -l --group-directories-first --dereference
total 0
drwxr-xr-x  1 nexion nexion      0 Aug 15 19:31 anime
drwxr-xr-x  1 nexion nexion      0 Aug 15 19:31 documents
drwxrwsr-x+ 1 nexion nexion   4902 Aug 13 16:27 downloads
drwxr-xr-x  1 nexion nexion      0 Aug 15 19:34 library
drwxrws---+ 1 nexion nexion   5328 Nov 19  2016 music
drwxr-xr-x  1 nexion nexion      0 Aug 15 19:35 something_else
-rw-r--r--  1 nexion nexion      0 Aug 15 19:32 av.conf
-rw-r--r--  1 nexion nexion      0 Aug 15 19:31 bookmarks.txt
prw-r--r--  1 nexion nexion      0 Aug 15 19:37 namedpipe
```

Listing directories and symlinked directories before the other files is also the default behaviour of Midnight Commander file manager.

I understand processing symlinks might not be desired by every user and somewhat slow down the sorting, but probably this can be optional?

**Comment by jarun** [comment]
> but probably this can be optional?

It affects several other things when you start making the changes... e.g., colors.

If you are interested, please raise a PR that adds a change to the view only - instead of showing:

    2018-08-15 21:42        @  testlink@

show

    2018-08-15 21:42        @d  testlink@d
The `d` after `@` would indicate that it's a link to a directory.

**Comment by jarun** [comment]
or even better `@/`

**Comment by jarun** [comment]
In the meantime, I have pushed a commit that indicates the link target is a dir at 75f4e032457fd9ce7f59882dff648e1f5333847a.

**Comment by jarun** [comment]
I have pushed c636db94bc62292c500095ee06e788b0eaaa835c to indicate the target is a dir in file stats.

**Comment by jarun** [comment]
Symbolic links to dirs are shown using indicator `@/` at 70df4fb3c20930aa97b1b2b593619d16aa9ebab1.

**Comment by jarun** [comment]
On Ubuntu, I keep seeing:

```
/usr/local/bin/nlay: 45: /usr/local/bin/nlay: Syntax error: "&&" unexpected
/usr/local/bin/nlay: 45: /usr/local/bin/nlay: Syntax error: "&&" unexpected
/usr/local/bin/nlay: 45: /usr/local/bin/nlay: Syntax error: "&&" unexpected
/usr/local/bin/nlay: 45: /usr/local/bin/nlay: Syntax error: "&&" unexpected
/usr/local/bin/nlay: 45: /usr/local/bin/nlay: Syntax error: "&&" unexpected
/usr/local/bin/nlay: 45: /usr/local/bin/nlay: Syntax error: "&&" unexpected
```

- `^/` does not work
- `vlock` is not invoked

I think we are better off using a bash script that _works correctly_.

One can have his own `nlay` in any preferred language. I don't have the bandwidth to get something working myself.

**Comment by iandhall** [comment]
When starting in Navigate-as-you-type mode, please could you add an option to disable the behaviour where nnn auto enters a directory when the filter matches a single directory? I would prefer to see a single filter result and then manually change to the directory if needed or just copy the dir without seeing the contents.

Please could selected files be visually marked? Would it be possible to set a different background colour or add a character to indicate which files have been selected?

**Comment by jarun** [comment]
> When starting in Navigate-as-you-type mode, please could you add an option to disable the behaviour where nnn auto enters a directory when the filter matches a single directory? I would prefer to see a single filter result and then manually change to the directory if needed or just copy the dir without seeing the contents.

I would throw in an env var to do that.

> Please could selected files be visually marked? Would it be possible to set a different background colour or add a character to indicate which files have been selected?

Those would be nasty hacks introducing additional checks at every row print. I understand your problem. I will explore other possibilities.

**Comment by jarun** [comment]
Option to disable dir auto-select pushed at e454078571f3a995cf1d8310b068a47d51a840d0.

**Comment by jarun** [comment]
I have added the second item to the ToDo list.

**Comment by jarun** [comment]
@iandhall copy buffer listing in multi-copy mode is implemented at 1562939325ce31c30a20a5bdd6d1262461e7f8fc.

On a separate note, if you believe the time and energy I spent in implementing these is worth it, please consider making a donation via the PayPal link on the homepage.

**Comment by mzanibelli** [comment]
Hi @jarun,

I wonder if there would be a way to tell `nnn` to print the path of the selected file on standard output and exit instead of triggering `xdg-open` or `$EDITOR` by default. I've been trying to hack it with something like `NNN_USE_EDITOR=1 VISUAL=echo nnn` but it obviously doesn't work very well as the program doesn't exit afterwards.

I'd love to be able to communicate with let's say Vim and use `nnn` as a file chooser leveraging the new `:terminal` feature. I currently use `ranger` which has a `--choosefile` option but it's very slow and I don't need so many features.
Scripts could be a way to do that but it doesn't run by default on Enter.  Any ideas based on clever environment variables use? If not, could you please consider adding this? I'd try but my understanding of C probably isn't good enough. 

Nice work by the way!

**Comment by jarun** [comment]
@mzanibelli This is interesting! I wasn't aware of the `:terminal` command. I will see what I can do.

However, I don't see much difference between the workflow you mentioned and the following (where `nnn` initiates the workflow instead of `vim`):

- run `nnn`
- navigate to a file and open it using `vim` (`NNN_USE_EDITOR` should be set with `EDITOR` set to `vim`)
- edit and close the file
- navigate to another file and open it with `nnn`
...

what difference do you see? (I'm sure I am missing something here.)

**Comment by mzanibelli** [comment]
Well, apart from my habit of starting my day with `vim` and ending it with `:quit`, the benefits I can think of are:

* the ability to choose from Vim which file selector to use (the `fzf` fuzzy finder works exactly the way I'm talking about, so it's a matter of replacing the name of the binary in the corresponding VimL code)

* starting Vim in the project root, and stay there, but mapping a key to somthing like `:terminal nnn %:h` to quickly navigate between files in the folder of the one being currently edited

* integrate with a lot more tools and scripts

**Comment by jarun** [comment]
> in the corresponding VimL code

Can you please share this? I need to understand what it is.

I have written a quick patch which we can extend to multiple files here: https://github.com/jarun/nnn/tree/nnn_print

See if it's the thing you need.

Also, I found a vim plugin here: https://github.com/ranger/ranger/blob/master/examples/vim_file_chooser.vim
which reads in the file names from a temp file. Are you referring to this script?

**Comment by jarun** [comment]
Little more detail on the patch:

The patch writes the selected file name to `stdout`.

Now, if you are referring to the vim plugin then it's _absolutely easy_ to implement it with `nnn`. We write selection names to a file `.nnncp`. Currently we remove the file when `nnn` exits. We can provide a hook to keep even after `nnn` quits and the plugin can simply read the file in.

**Comment by jarun** [comment]
@mzanibelli Yes, on closer inspection, I think we are discussing the same script. So I can throw in a new option that takes in a custom `.nnncp` file and _doesn't remove_ the file on exit. So you just select a file (or multiple files) and exit `nnn` to get the list and do whatever you want with it.

Sounds good?

**Comment by mzanibelli** [comment]
My VimL code for reference: https://github.com/mzanibelli/dotvim/blob/6dd7012a973bbc7711070dbb2f9487a87f5b485e/autoload/edit.vim#L10
Goal: replace l11 with `nnn`

We're getting there! I was not aware of `.nnncp` which seems even closer to what `ranger` does with `--choosefile`. Quick thoughts:

* `stdout` is better because it is, in my opinion, the way to go in any *nix environment

* I see that the only required code was to use `mkpath()` before printing so I think I can do exactly what I want now, I'll try to submit a patch

* bonus: print path and quit on`→, ↵, l, ^M` when using this feature

* mutiple file selection would be nice indeed!

**Comment by jarun** [comment]
> `stdout` is better because it is, in my opinion, the way to go in any *nix environment

See my last comment. In that case, we can have the option such that, if no filepath is provided, `nnn` writes the selection list to `stdout` before exit.

**Comment by jarun** [comment]
I will reopen the issue and copy these notes in that. We need to have the context for this discussion.

**Comment by jarun** [comment]
Rolled at #137.

**Comment by jarun** [comment]
@Nexion I had a hard time finding your request to place symlinked dirs along with dirs but it is now implemented at commit 725349976a2a5144f34d7bd7332444a53bbf81ac.
